src/routes/teams.rs

```diff
@@ -1,25 +1,227 @@
+use crate::auth::get_user_from_headers;
 use crate::database::models::TeamMember;
-use crate::models::teams::TeamId;
+use crate::models::teams::{Permissions, TeamId};
+use crate::models::users::UserId;
 use crate::routes::ApiError;
-use actix_web::{get, web, HttpResponse};
+use actix_web::{delete, get, patch, post, web, HttpRequest, HttpResponse};
+use serde::{Deserialize, Serialize};
 use sqlx::PgPool;
 
 #[get("{id}/members")]
 pub async fn team_members_get(
+    req: HttpRequest,
     info: web::Path<(TeamId,)>,
     pool: web::Data<PgPool>,
 ) -> Result<HttpResponse, ApiError> {
     let id = info.into_inner().0;
     let members_data = TeamMember::get_from_team(id.into(), &**pool).await?;
 
+    let current_user = get_user_from_headers(req.headers(), &**pool).await.ok();
+
+    if let Some(user) = current_user {
+        let team_member = TeamMember::get_from_user_id(id.into(), user.id.into(), &**pool)
+            .await
+            .map_err(|e| ApiError::DatabaseError(e.into()))?;
+
+        if team_member.is_some() {
+            let team_members: Vec<crate::models::teams::TeamMember> = members_data
+                .into_iter()
+                .map(|data| crate::models::teams::TeamMember {
+                    user_id: data.user_id.into(),
+                    name: data.name,
+                    role: data.role,
+                    permissions: data.permissions,
+                })
+                .collect();
+
+            return Ok(HttpResponse::Ok().json(team_members));
+        }
+    }
+
     let team_members: Vec<crate::models::teams::TeamMember> = members_data
         .into_iter()
         .map(|data| crate::models::teams::TeamMember {
             user_id: data.user_id.into(),
             name: data.name,
             role: data.role,
+            permissions: Permissions::default(),
         })
         .collect();
 
     Ok(HttpResponse::Ok().json(team_members))
 }
+
+#[post("{id}/join")]
+pub async fn join_team(
+    req: HttpRequest,
+    info: web::Path<(TeamId,)>,
+    pool: web::Data<PgPool>,
+) -> Result<HttpResponse, ApiError> {
+    let team_id = info.into_inner().0.into();
+    let current_user = get_user_from_headers(req.headers(), &**pool)
+        .await
+        .map_err(|_| ApiError::AuthenticationError)?;
+
+    // Edit Team Member to set Accepted to True
+    TeamMember::edit_team_member(
+        team_id,
+        current_user.id.into(),
+        None,
+        None,
+        Some(true),
+        &**pool,
+    )
+    .await?;
+
+    Ok(HttpResponse::Ok().body(""))
+}
+
+#[post("{id}/members")]
+pub async fn add_team_member(
+    req: HttpRequest,
+    info: web::Path<(TeamId,)>,
+    pool: web::Data<PgPool>,
+    new_member: web::Json<crate::models::teams::TeamMember>,
```

Comment:
This should have a separate struct for creating a member, with most fields being optional.  `permissions` and `role` should be optional and have reasonable defaults, and `name` (if you even allow it to be specified) should default to the new user's username.